src/arch/mips/lib/debug/stack_iter.c

```diff
@@ -0,0 +1,82 @@
+/**
+ * @brief
+ * @author Nastya Nizharadze
+ * @date 30.07.2019
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include <hal/context.h>
+#include <util/lang.h>
+#include <util/math.h>
+#include <assert.h>
+#include "stack_iter.h"
+#include "debug/symbol.h"
+
+/* Mips stack:
+ *
+ * low address ^
+ * 
+ *  -------- frame begining
+ * |        | ---		<- new fp + offset
+ * |   ra   |  ^		<- new fp + offset - 4
+ * | old fp |  |		<- new fp + offset - 8
+ * |   ...  |  | offset
+ * |        |  |
+ * |        |  v
+ * |        | ---		<- new fp
+ *  -------- frame ending
+ * 
+ *  */
+int offset (void *ra) {
+	uint32_t opcode;
+	const struct symbol *a;
+
+/* Find an address of first function's instruction */
+	a = symbol_lookup(ra);
+	if (a == NULL) {
+		return 0;
+	}
+/* Find an operation code. it is always addiu sp,sp,x 
+ * where x is an offset */
+	opcode = *((unsigned *) (a->addr));
+/* The operation code for addiu sp,sp,x should always started whith 0X27bd */
+	assert((opcode & 0xffff0000) == 0x27bd0000);
+/* The last 4 digits are for offset*/
+	return (0x10000 - (opcode & 0xffff));
+}
+
+int stack_iter_next(stack_iter_t *f) {
+	uint32_t offset_cur;
+
+	if (!(offset_cur = offset(f->ra))) {
+		return 0;
+	}
+/* To get an old fp we need to cast the contents of this  |
+ * block of memory to the void pointer type				  V */
+	f->fp = (void *) *((unsigned *) (f->fp + offset_cur - 8));
+	f->ra = (void *) *((unsigned *) (f->fp - 4));
+/* When the old fp become 0 stack is ower */
+	return (void *) *((unsigned *) (f->fp + offset_cur - 8)) != 0;
```

Comment:
What happens here? You already updated `f->fp` to the new value above:

```
f->fp = (void *) *((unsigned *) (f->fp + offset_cur - 8));
```

And at this line you increment `fp` at the same value once more in order to compare with 0.

Reply:
It is supposed to be a next fp. Done.